sop-core/src/Data/SOP/NP.hs

```diff
@@ -1,8 +1,13 @@
-{-# LANGUAGE PolyKinds, StandaloneDeriving, UndecidableInstances #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+
 -- | n-ary products (and products of products)
 module Data.SOP.NP
   ( -- * Datatypes
-    NP(..)
+    NP(.., Nil, (:*))
```

Comment:
Do we want to export the `NP` constructor, or do we want to move it to an `Internal` module?

Reply:
Moving it would make sense to me, otherwise there are bound to be uses of `NP` that use the patterns dictionary and they would now suddenly get much worse efficiency.

Reply:
I would definitely want to be able to access the NP constructor as a developer, but putting it in a `.Internal` modules makes sense as it doesn't provide a safe interface, which `Nil` and `(:*)` do. If I _know what I'm doing_ and performance matters to me, I'd want to be able to work with the Vector directly and convince myself I did so safely.